src/asynch/wifi.rs

```diff
@@ -0,0 +1,308 @@
+    /// Error when receiving local address information
+    type AddressError: Debug;
+
+    /// Errors for configuration commands
+    type ConfigurationErrors: Debug;
+
+    /// Errors when restarting the module
+    type RestartError: Debug;
+
+    /// Connects to an WIFI access point and returns the connection state
+    fn join(&mut self, ssid: &str, key: &str) -> impl core::future::Future<Output = Result<JoinState, Self::JoinError>>;
+
+    /// Returns the current WIFI connection status
+    fn get_join_status(&mut self) -> impl core::future::Future<Output = JoinState>;
+
+    /// Returns local address information
+    fn get_address(&mut self) -> impl core::future::Future<Output = Result<LocalAddress, Self::AddressError>>;
+
+    /// Enables/Disables auto connect, so that ESP-AT whether automatically joins to the stored AP when powered on.
+    fn set_auto_connect(&mut self, enabled: bool) -> impl core::future::Future<Output = Result<(), Self::ConfigurationErrors>>;
+
+    /// Restarts the module and blocks until ready
+    fn restart(&mut self) -> impl core::future::Future<Output = Result<(), Self::RestartError>>;
+}
+
+pub struct InnerAdapter<
+    'urc_sub,
+    A: AtatClient,
+    const TX_SIZE: usize,
+    const RX_SIZE: usize,
+    const URC_CAPACITY: usize,
+> {
+    /// ATAT client
+    pub(crate) client: A,
+
+    /// URC message subscriber
+    pub(crate) urc_subscription: UrcSubscription<'urc_sub, URCMessages<RX_SIZE>, URC_CAPACITY, 1>,
+
+    /// Network state
+    pub(crate) session: Session<RX_SIZE>,
+
+    /// Timeout for data transmission
+    pub(crate) send_timeout: Duration,
+}
+
+impl<
+        'urc_sub,
+        A: AtatClient,
+        const TX_SIZE: usize,
+        const RX_SIZE: usize,
+        const URC_CAPACITY: usize,
+    > InnerAdapter<'urc_sub, A, TX_SIZE, RX_SIZE, URC_CAPACITY>
+{
+    pub fn new(
+        client: A,
+        urc_subscription: UrcSubscription<'urc_sub, URCMessages<RX_SIZE>, URC_CAPACITY, 1>,
+        send_timeout: Duration,
+    ) -> Self {
+        Self {
+            client,
+            urc_subscription,
+            session: Session::default(),
+            send_timeout,
+        }
+    }
+
+    pub(crate) fn process_urc_messages(&mut self) {
+        while let Some(message) = self.urc_subscription.try_next_message_pure() {
+            self.session.handle_urc(message)
+        }
+    }
+
+    async fn set_station_mode(&mut self) -> Result<(), JoinError> {
+        let command = WifiModeCommand::station_mode();
+        self.send_command(command).await?;
+
+        Ok(())
+    }
+
+    async fn connect_access_point(&mut self, ssid: &str, key: &str) -> Result<(), JoinError> {
+        if ssid.len() > 32 {
+            return Err(JoinError::InvalidSSDLength);
+        }
+
+        if key.len() > 63 {
+            return Err(JoinError::InvalidPasswordLength);
+        }
+
+        let command = AccessPointConnectCommand::new(String::from_str(ssid).unwrap(), String::from_str(key).unwrap());

```

Comment:
As the logic is identical to the sync. part. What about uniforming it?
Maybe as some sort of `AccessPointConnectCommand::from_cred(ssid, key)?`?